### PR TITLE
chainset::quantiles function catches exceptions raised by stan::math::quantiles when sample contains NaN's

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [ ] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

For samples where one or more draws is NaN, function `stan::math::quantiles` throws a `std::invalid_argument` exception.   Modify `quantile` and `quantiles` functions so that they catch this, as well as the `std::domain_error` exception (thrown if the requested probability is outside of interval [0, 1]) and return `NaN`.

#### Intended Effect

Prevent runtime exceptions being thrown by `stansummary` function.

#### How to Verify

Unit tests.

#### Side Effects

N/A

#### Documentation

N/A

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by WardBrian** [comment]
It seems like the stan csv reader is using istream’s operator>> to parse numbers, which as far as I can tell is not defined for nan or infinite values. I’m guessing the reason we were witnessing something on MacOS only is that the different platforms do different things in this case.

so:
- this change is good, but the test might need to be different
- csv_reader should probably use std::stod, rather than a stringstream and >>?

**Comment by mitzimorris** [comment]
It looks like Ubuntu doesn't throw these exceptions.
Will revise PR accordingly.

**Comment by WardBrian** [comment]
It's because ubuntu is reading the `nan` as 0, so it's not throwing an exception:
https://godbolt.org/z/b6G336qYP

**Comment by WardBrian** [comment]
I suspect changing 
https://github.com/stan-dev/stan/blob/595ae416900093bdaa1309634c927a4814038b49/src/stan/io/stan_csv_reader.hpp#L337

to `samples(row, col) = std::stod(line);` will both make the test here work and fix a whole other set of bugs-lurking-in-the-shadows. Right now if some of the draws are nan, they're just being treated as zero (possibly in a platform-dependent way).

Not that this is performance critical, but I'd also think it would be faster (though there is a lot about the current code in that function that concerns me in _that_ regard)

**Comment by mitzimorris** [comment]
just seeing this now.   will change the CSV reader.

**Comment by mitzimorris** [comment]
> to `samples(row, col) = std::stod(line);` will both make the test here work and fix a whole other set of bugs-lurking-in-the-shadows. Right now if some of the draws are nan, they're just being treated as zero (possibly in a platform-dependent way).

one of the eight schools test files is triggering  a std::out_of_range error from std::stod.   indeed, the value is very close to zero - for some reason, there's an `accept_stat` value:  `6.77055e-309`.

**Comment by WardBrian** [comment]
> from std::stod. indeed, the value is very close to zero

it seems using stold is able to parse that value, and then down casting to a double should work. The csv files were originally produced from double values, so I’m not worried about the down casting being lossy

**Comment by WardBrian** [comment]
If the csv reader changes do work, I slightly preferred the rendition that caught logic_error to the current code that does a check that the quantiles function then repeats internally

File: src/stan/mcmc/chainset.hpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include <stan/math/prim.hpp>
 #include <stan/math/prim/fun/constants.hpp>
 #include <stan/math/prim/fun/quantile.hpp>
+#include <stan/analyze/mcmc/check_chains.hpp>
```
Comment (WardBrian) [inline src/stan/mcmc/chainset.hpp]: Include not needed

File: src/stan/io/stan_csv_reader.hpp [hunk 2]
```diff
@@ -334,7 +334,12 @@ class stan_csv_reader {
         for (int col = 0; col < cols; col++) {
           std::getline(ls, line, ',');
           boost::trim(line);
-          std::stringstream(line) >> samples(row, col);
+          try {
+            auto raw = std::stold(line);
```
Comment (WardBrian) [inline src/stan/io/stan_csv_reader.hpp]: thoughts on doing this in one line and not needing this temporary?

**Review by WardBrian — APPROVED** [review]
Two nits but otherwise looks good!
Thanks so much for tracking this down!

**Comment by mitzimorris** [comment]
changes made - I leave it to you to merge.

**Comment by WardBrian** [comment]
Many thanks!

**Pull request merged**